src/layers/domain/core/product.py

```diff
@@ -1,98 +1,132 @@
+    ACCREDITED_SYSTEM_ID = "ACCREDITED_SYSTEM_ID"
 
-        super().__init__(id=id, name=name)
 
-    def create_product_team(
-        self, id: str, name: str, owner: User
-    ) -> tuple["ProductTeam", "ProductTeamCreatedEvent"]:
-        """
-        Creates an instance of a ProductTeam
-        """
-        product_team = ProductTeam(id=id, name=name, organisation=self, owner=owner)
-        event = ProductTeamCreatedEvent(product_team=product_team)
-        return (product_team, event)
+class ProductKey(BaseModel):
+    type: ProductKeyType = Field()
 
 
-class ProductTeam(Entity[UUID]):
-    """
-    A ProductTeam is created from a Supplier OdsOrganisation, and is the entity
-    that creates and owns Products.
-    """
+class ProductType(Enum):
+    SERVICE = "SERVICE"
+    API = "API"
 
-    def __init__(self, id: UUID, name: str, organisation: OdsOrganisation, owner: User):
-        validate_uuid(uuid=id)
-        validate_type(obj=organisation, expected_type=OdsOrganisation)
-        validate_type(obj=owner, expected_type=User)
 
-        super().__init__(id=id, name=name)
-        self.organisation = organisation
-        self.owner = owner
+class ProductStatus(Enum):
+    ACTIVE = "ACTIVE"
+    INACTIVE = "INACTIVE"
 
-    def create_product(self) -> "Product":
-        """
-        Creates an instance of a new Product
-        """
-        pass
 
+class Product(AggregateRoot):
+    id: UUID = Field()
+    name: str = Field(pattern="^[a-z][a-z0-9_]+$")
+    type: ProductType = Field()
+    status: ProductStatus = Field(default=ProductStatus.ACTIVE)
+    product_team_id: UUID = Field()
+    ods_code: str = Field()
+    relationships: dict[UUID, Relationship] = (lambda: {})()
+    keys: dict[str, ProductKey] = (lambda: {})()
 
-class Product(Entity[UUID], QuestionnaireEntity):
-    """
-    A Product represents logical and physical software products.  A Product may
-    be referenced by multiple keys, such as internal GUID, Product Id, ASID,
-    etc.
-    """
+    def rename(self, name: str):
+        raise NotImplementedError()
 
-    def __init__(
-        self,
-        id: str,
-        name: str,
-        questionnaires: set[str] = None,
-        dependency_questionnaires: set[str] = None,
-    ):
-        validate_product_id_or_asid(id=id)
+    def add_relationship(
+        self, target: "Product", type: RelationshipType
+    ) -> RelationshipAddedEvent:
+        self.relationships[target.id] = Relationship(type=type)
+        event = RelationshipAddedEvent(id=self.id, target=target.id, type=type)
+        return self.add_event(event=event)
 
-        super().__init__(id=id, name=name)
-        self._questionnaires = default_set(questionnaires)
-        self._dependency_questionnaires = default_set(dependency_questionnaires)
+    def remove_relationship(self, target: str) -> RelationshipRemovedEvent:
+        target_id = target.id if isinstance(target, Product) else target
+        del self.relationships[target_id]
```

Comment:
will need to wrap this up in a custom exception, otherwise will throw 500 when key doesn't exist